starboard/linux/shared/pre_app_recommendation_service.cc

```diff
@@ -0,0 +1,183 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/linux/shared/pre_app_recommendation_service.h"
+
+#include <memory>
+#include <string>
+
+#include "starboard/common/log.h"
+#include "starboard/common/string.h"
+#include "starboard/configuration.h"
+#include "starboard/linux/shared/platform_service.h"
+#include "starboard/shared/starboard/application.h"
+
+// Omit namespace linux due to symbol name conflict.
+namespace starboard {
+namespace shared {
+namespace {
+typedef struct PreAppRecommendationsPlatformServiceImpl
+    : public PlatformServiceImpl {
+  // Define additional data field.
+  // variable_1, variable_2,...
+  PreAppRecommendationsPlatformServiceImpl(
+      void* context,
+      ReceiveMessageCallback receive_callback)
+      : PlatformServiceImpl(context, receive_callback) {}
+
+  // Default constructor.
+  PreAppRecommendationsPlatformServiceImpl() = default;
+
+} PreAppRecommendationsPlatformServiceImpl;
+
+// Use HTTP status code in response to YouTube application's method recommend
+// call.
+const char kSuccess[] = "200";
+const char kBadRequest[] = "400";
+
+bool Has(const char* name) {
+  // Check if platform has service name.
+  return strcmp(name, kPreappRecommendationServiceName) == 0;
+}
+
+PlatformServiceImpl* Open(void* context,
+                          ReceiveMessageCallback receive_callback) {
+  SB_DCHECK(context);
+  SB_LOG(INFO) << "Open() service created: "
+               << kPreappRecommendationServiceName;
+
+  return new PreAppRecommendationsPlatformServiceImpl(context,
+                                                      receive_callback);
+}
+
+void Close(PlatformServiceImpl* service) {
+  // Function Close shouldn't manually delete PlatformServiceImpl pointer,
+  // because it is managed by unique_ptr in Platform Service.
+  SB_LOG(INFO) << "Perform actions before gracefully shutting down the service";
```

Comment:
This log line should probably say something like `SB_LOG(INFO) << kPreappRecommendationServiceName << " closed"`